sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java

```diff
@@ -174,28 +174,39 @@ static boolean hasOnlyJavaUdfInProjects(RelOptRuleCall x) {
           } else if (rexNode instanceof RexLiteral) {
             SqlTypeName typeName = ((RexLiteral) rexNode).getTypeName();
             switch (typeName) {
-              case NULL:
+              case DOUBLE:
               case BOOLEAN:
-              case CHAR:
-              case BINARY:
+              case VARCHAR:
+              case VARBINARY:
               case DECIMAL:
+              case TIME:
+              case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+              case TIMESTAMP:
+              case ARRAY:
+              case ROW:
                 break;
+              case BIGINT: // BEAM-11989
+              case DATE: // BEAM-11990
               default:
-                // Reject unsupported literals
                 return false;
             }
           } else if (rexNode instanceof RexInputRef) {
             SqlTypeName typeName = ((RexInputRef) rexNode).getType().getSqlTypeName();
             switch (typeName) {
-              case BIGINT:
               case DOUBLE:
               case BOOLEAN:
               case VARCHAR:
               case VARBINARY:
               case DECIMAL:
+              case TIME:
+              case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+              case TIMESTAMP:
+              case ARRAY:
+              case ROW:
```

Comment:
I believe you need to recurse into `((RexInputRef) rexNode).getType().getFieldList()` to verify the struct only contains supported types.

Reply:
done